tests/test_docs_extraction.py

```diff
@@ -0,0 +1,139 @@
+from typing_extensions import Annotated, TypedDict
+
+from pydantic import BaseModel, ConfigDict, Field, TypeAdapter
+from pydantic.dataclasses import dataclass as pydantic_dataclass
+
+
+def test_model_no_docs_extraction():
+    class Model(BaseModel):
+        a: int = 1
+        """A docs"""
+
+        b: str = '1'
+
+        """B docs"""
+
+    assert Model.model_fields['a'].description is None
+    assert Model.model_fields['b'].description is None
+
+
+def test_model_docs_extraction():
+    class Model(BaseModel):
```

Comment:
Unfortunately `inspect.getsource` isn't working as expected with classes. Tests are failing under 3.9 because prior to https://github.com/python/cpython/pull/10307, the following would happen:

<details><summary>MRE</summary>

```python
import inspect


def test1():
    class Model:
        a = 1

    print(inspect.getsource(Model))


def test2():
    class Model:
        a = 2

    print(inspect.getsource(Model))


test1()
#> uses Model with a = 1
test2()
#> uses Model with a = 1 (wrong!)
```
</details>

This has been fixed since then, but encountered this one: https://github.com/python/cpython/issues/106727

I think this isn't a big deal, as users will probably not defined two models with the same name in the same module.

Reply:
That kinda sounds like a dealbreaker to me. I don’t know why you think it’s not common, it seems to me like it would happen all the time in tests.

Reply:
Imo it's not common from a user perspective (as this only occurs if in the same module). But it is indeed way more common in tests 😕

Reply:
Users write tests…

Reply:
Ah yes didn't think about that 😅 That was what I was afraid of by using `inspect`, and I don't see any other path except shipping our own source parsing logic, which probably isn't a good idea

Reply:
I'll try to improve class detection on the Python side (see the related issue on cpython), but unfortunately it will not be able to cover all edge cases, such as:

```python
if cond:
    class A: pass
else:
    class A: pass        
```

As this feature would be off by default, users could be warned about the potential unexpected behavior in the documentation. I'll let you decide on the output of this PR

Reply:
So Python got some improvements since then:
- https://github.com/python/cpython/pull/106815
- https://github.com/python/cpython/pull/106968

It works way better if a method is defined on the class (thanks to `__code__.co_firstlineno`), but can still pick the wrong class otherwise.

Reply:
Am I understanding correctly that this misbehavior is only on Python<=3.9?

Also, when it gets the wrong source, does it lead to actual errors, or does it just retrieve the docstrings incorrectly?

If it only affects <=3.9 and fails gracefully in the sense that it just doesn't get the docstrings correctly, then unless I'm missing something I think the issues with `inspect.getsource` maybe shouldn't be a dealbreaker since it would require all of the following conditions to be met for it to cause problems:

* You are using Python<=3.9
* You have opted into the new config setting for this functionality
* You have defined a model with the same name multiple times in the same module
* You require the description attribute for the `FieldInfo` must be correct
* You aren't happy to (or aren't aware that you can) explicitly specify the description in `Field(...)` for affected attributes

In particular, for the case of writing tests that you mentioned @adriangb I would think that this would very rarely cause problems for tests that aren't explicitly testing this functionality.

If nothing else, it seems reasonable to me to emit a warning if this particular config flag is set in `Python<=3.9`.

(That said, it looks to me like there are other test failures happening right now, not just caused by this issue.)

Reply:
The following PR:
- https://github.com/python/cpython/pull/10307

was introduced in Python 3.9, and fixes class detection when they are nested in other classes.

Up until now, it would still fail with the following:

```python
import inspect

class Model:
    a = 1

class Model:
    a = 2

print(Model.a)
#> 2
print(inspect.getsource(Model))
#> class Model:
#>     a = 1
```

This got improved with the following PR (it will now pick the last class definition):
- https://github.com/python/cpython/pull/106815

But still can get it wrong if you're doing something like:

```python
if False:
    class Model:
        a = 1
else:
    class Model:
        a = 2
# Which really is a uncommon edge case
```

> when it gets the wrong source, does it lead to actual exceptions being raised, or does it just retrieve the docstrings incorrectly?

It will not raise currently, as applying field docstrings in a safe way, but I could maybe raise if we encounter unexpected fields (I'll have to see if it's doable when inheritance is being used).

So to summarize, all of your conditions should indeed be met, except:
- You are using Python<=3.9
- it fails gracefully (can be fixed as I said)

I'll be off for a while and will continue working on this in September if this is still accepted